File: app/src/main/java/com/junkiedan/junkietuner/core/PreferencesDataStoreHandler.java
package com.junkiedan.junkietuner.core;

import android.content.Context;
import androidx.datastore.preferences.core.MutablePreferences;
import androidx.datastore.preferences.core.Preferences;
import androidx.datastore.preferences.core.PreferencesKeys;
import androidx.datastore.preferences.rxjava3.RxPreferenceDataStoreBuilder;
import androidx.datastore.rxjava3.RxDataStore;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Single;

/**
 * Class contains reference to the Preferences DataStore data storage. For more information check
 * this <a href="https://developer.android.com/topic/libraries/architecture/datastore">link</a>
 * to the Android documentation. All methods are static inside this class (Therefore constructor
 * is empty and no instance of this class can be instantiated
 *
 * @author Stavros Barousis
 */
public class PreferencesDataStoreHandler {

    private static RxDataStore<Preferences> dataStore = null;
    private final static String DS_FILE_NAME = "settings";
    private final static Preferences.Key<Integer> CURRENT_TUNING_ID =
            PreferencesKeys.intKey("current_tuning_id");
    private final static Preferences.Key<Boolean> HAS_BEEN_INITIALIZED =
            PreferencesKeys.booleanKey("has_been_initialized");
    private final static Preferences.Key<Boolean> IS_TUNER_LOCKED =
            PreferencesKeys.booleanKey("is_tuner_locked");
    private final static Preferences.Key<Boolean> IS_LOAD_LAST_MUTED_STATE =
            PreferencesKeys.booleanKey("is_load_last_muted_state");
    private final static Preferences.Key<Boolean> IS_TUNING =
            PreferencesKeys.booleanKey("is_tuning");


    private PreferencesDataStoreHandler() {
        // Empty private constructor
    }

    /**
     * Datastore reference initializer (needs to be private)
     * @param context Current context that is passed as input
     */
    private static void initDataStore(Context context) {
        dataStore = new RxPreferenceDataStoreBuilder(context, DS_FILE_NAME).build();
    }

    /**
     * Retrieves data from the Preferences DataStore
     * @param context Current application context
     * @return Flowable object of the boolean value HAS_BEEN_INITIALIZED
     */
    public static Flowable<Boolean> hasBeenInitialized(Context context) {
        if (dataStore == null) {
            initDataStore(context);
        }
        return dataStore.data().map(preferences ->
                preferences.get(HAS_BEEN_INITIALIZED));
    }

    /**
     * Writes data to the Preferences DataStore
     * @param context Current application context
     * @param value Boolean value that will be stored on flag HAS_BEEN_INITIALIZED
     */
    public static void setHasBeenInitialized(Context context, boolean value) {
        if (dataStore == null) {
            initDataStore(context);
        }
        dataStore.updateDataAsync(prefsIn -> {
            MutablePreferences mutablePreferences = prefsIn.toMutablePreferences();
            mutablePreferences.set(HAS_BEEN_INITIALIZED, value);
            return Single.just(mutablePreferences);
        });
    }

    /**
     * Retrieves data from the Preferences DataStore
     * @param context Current application context
     * @return Flowable object of the integer value CURRENT_TUNING_ID
     */
    public static Flowable<Integer> getCurrentTuningId(Context context) {
        if (dataStore == null) {
            initDataStore(context);
        }
        return dataStore.data().map(preferences ->
                preferences.get(CURRENT_TUNING_ID));
    }

    /**
     * Writes data to the Preferences DataStore
     * @param context Current application context
     * @param newId Integer value that will be stored in CURRENT_TUNING_ID
     */
    public static void setCurrentTuningId(Context context, int newId) {
        if (dataStore == null) {
            initDataStore(context);
        }
        dataStore.updateDataAsync(prefsIn -> {
            MutablePreferences mutablePreferences = prefsIn.toMutablePreferences();
            mutablePreferences.set(CURRENT_TUNING_ID, newId);
            return Single.just(mutablePreferences);
        });
    }

    /**
     * Retrieves data from the Preferences DataStore
     * @param context Current application context
     * @return Flowable object of the boolean value IS_TUNER_LOCKED
     */
    public static Flowable<Boolean> getIsTunerLocked(Context context) {
        if (dataStore == null) {
            initDataStore(context);
        }
        return dataStore.data().map(preferences ->
                preferences.get(IS_TUNER_LOCKED));
    }

    /**
     * Writes data to the Preferences DataStore
     * @param context Current application context
     * @param value Boolean value that will be stored on flag IS_TUNER_LOCKED
     */
    public static void setIsTunerLocked(Context context, boolean value) {
        if (dataStore == null) {
            initDataStore(context);
        }
        dataStore.updateDataAsync(prefsIn -> {
            MutablePreferences mutablePreferences = prefsIn.toMutablePreferences();
            mutablePreferences.set(IS_TUNER_LOCKED, value);
            return Single.just(mutablePreferences);
        });
    }

    /**
     * Retrieves data from the Preferences DataStore
     * @param context Current application context
     * @return Flowable object of the boolean value IS_LOAD_LAST_MUTED_STATE
     */
    public static Flowable<Boolean> getIsLoadLastMutedState(Context context) {
        if (dataStore == null) {
            initDataStore(context);
        }
        return dataStore.data().map(preferences ->
                preferences.get(IS_LOAD_LAST_MUTED_STATE));
    }

    /**
     * Writes data to the Preferences DataStore
     * @param context Current application context
     * @param value Boolean value that will be stored on flag IS_LOAD_LAST_MUTED_STATE
     */
    public static void setIsLoadLastMutedState(Context context, boolean value) {
        if (dataStore == null) {
            initDataStore(context);
        }
        dataStore.updateDataAsync(prefsIn -> {
            MutablePreferences mutablePreferences = prefsIn.toMutablePreferences();
            mutablePreferences.set(IS_LOAD_LAST_MUTED_STATE, value);
            return Single.just(mutablePreferences);
        });
    }

    /**
     * Retrieves data from the Preferences DataStore
     * @param context Current application context
     * @return Flowable object of the boolean value IS_TUNING
     */
    public static Flowable<Boolean> getIsTuning(Context context) {
        if (dataStore == null) {
            initDataStore(context);
        }
        return dataStore.data().map(preferences ->
                preferences.get(IS_TUNING));
    }

    /**
     * Writes data to the Preferences DataStore
     * @param context Current application context
     * @param value Boolean value that will be stored on flag IS_TUNING
     */
    public static void setIsTuning(Context context, boolean value) {
        if (dataStore == null) {
            initDataStore(context);
        }
        dataStore.updateDataAsync(prefsIn -> {
            MutablePreferences mutablePreferences = prefsIn.toMutablePreferences();
            mutablePreferences.set(IS_TUNING, value);
            return Single.just(mutablePreferences);
        });
    }
}

File: app/src/main/java/dev/thestbar/tunify/core/PreferencesDataStoreHandler.kt
package dev.thestbar.tunify.core

import android.content.Context
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dev.thestbar.tunify.data.preferences.ThemePreference
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

private val Context.dataStore by preferencesDataStore(name = "settings")

object PreferencesDataStoreHandler {

    private val CURRENT_TUNING_ID = intPreferencesKey("current_tuning_id")
    private val HAS_BEEN_INITIALIZED = booleanPreferencesKey("has_been_initialized")
    private val IS_TUNER_LOCKED = booleanPreferencesKey("is_tuner_locked")
    private val IS_LOAD_LAST_MUTED_STATE = booleanPreferencesKey("is_load_last_muted_state")
    private val IS_TUNING = booleanPreferencesKey("is_tuning")
    private val THEME_PREFERENCE = stringPreferencesKey("theme_preference")

    fun hasBeenInitialized(context: Context): Flow<Boolean?> =
        context.dataStore.data.map { it[HAS_BEEN_INITIALIZED] }

    suspend fun setHasBeenInitialized(context: Context, value: Boolean) {
        context.dataStore.edit { it[HAS_BEEN_INITIALIZED] = value }
    }

    fun getCurrentTuningId(context: Context): Flow<Int?> =
        context.dataStore.data.map { it[CURRENT_TUNING_ID] }

    suspend fun setCurrentTuningId(context: Context, newId: Int) {
        context.dataStore.edit { it[CURRENT_TUNING_ID] = newId }
    }

    fun getIsTunerLocked(context: Context): Flow<Boolean?> =
        context.dataStore.data.map { it[IS_TUNER_LOCKED] }

    suspend fun setIsTunerLocked(context: Context, value: Boolean) {
        context.dataStore.edit { it[IS_TUNER_LOCKED] = value }
    }

    fun getIsLoadLastMutedState(context: Context): Flow<Boolean?> =
        context.dataStore.data.map { it[IS_LOAD_LAST_MUTED_STATE] }

    suspend fun setIsLoadLastMutedState(context: Context, value: Boolean) {
        context.dataStore.edit { it[IS_LOAD_LAST_MUTED_STATE] = value }
    }

    fun getIsTuning(context: Context): Flow<Boolean?> =
        context.dataStore.data.map { it[IS_TUNING] }

    suspend fun setIsTuning(context: Context, value: Boolean) {
        context.dataStore.edit { it[IS_TUNING] = value }
    }

    fun getThemePreference(context: Context): Flow<ThemePreference> =
        context.dataStore.data.map { prefs ->
            when (prefs[THEME_PREFERENCE]) {
                "LIGHT" -> ThemePreference.LIGHT
                "DARK"  -> ThemePreference.DARK
                else    -> ThemePreference.SYSTEM
            }
        }

    suspend fun setThemePreference(context: Context, value: ThemePreference) {
        context.dataStore.edit { it[THEME_PREFERENCE] = value.name }
    }
}
